File: src/Shared/aes.m
//
//  aes.m
//  Go Map!!
//
//  Created by Bryce on 3/2/20.
//  Copyright © 2020 Bryce. All rights reserved.
//

#import <CommonCrypto/CommonCryptor.h>

#import "aes.h"

@implementation aes

+ (NSData *)encryptData:(NSData *)data key:(const uint8_t *)key
{
    return [self aesOperation:kCCEncrypt OnData:data key:key];
}

+ (NSData *)decryptData:(NSData *)data key:(const uint8_t *)key
{
    return [self aesOperation:kCCDecrypt OnData:data key:key];
}

+ (NSData *)aesOperation:(CCOperation)op
                  OnData:(NSData *)data
                     key:(const uint8_t *)key
{
    uint8_t buffer[ data.length + kCCKeySizeAES128 ];
    size_t bufferLen;
    CCCrypt(op,
            kCCAlgorithmAES128,
            kCCOptionPKCS7Padding,
            key,
            kCCKeySizeAES128,
            NULL,
            data.bytes,
            data.length,
            buffer,
            sizeof buffer,
            &bufferLen);
    return [NSData dataWithBytes:buffer length:bufferLen];
}

static const uint8_t key[] = { 250, 157, 60, 79, 142, 134, 229, 129, 138, 126, 210, 129, 29, 71, 160, 208 };

+(NSString *)encryptString:(NSString *)string
{
	NSData * data = [string dataUsingEncoding:NSUTF8StringEncoding];
	NSData * dec = [aes encryptData:data key:key];
	return [dec base64EncodedStringWithOptions:0];
}

+(NSString *)decryptString:(NSString *)string
{
	NSData * data = [[NSData alloc] initWithBase64EncodedString:string options:0];
	NSData * dec = [aes decryptData:data key:key];
	return [[NSString alloc] initWithData:dec encoding:NSUTF8StringEncoding];
}

@end

File: src/Shared/aes.swift
//  Converted to Swift 5.4 by Swiftify v5.4.27034 - https://swiftify.com/
//
//  aes.swift
//  Go Map!!
//
//  Created by Bryce on 3/2/20.
//  Copyright © 2020 Bryce. All rights reserved.
//

import CommonCrypto
import Foundation

let key: [__uint8_t] = [250, 157, 60, 79, 142, 134, 229, 129, 138, 126, 210, 129, 29, 71, 160, 208]

@objcMembers
class aes: NSObject {
    class func encryptData(_ data: Data, key: UnsafePointer<UInt8>) -> Data {
        return self.aesOperation(CCOperation(kCCEncrypt), on: data, key: key)
    }

    class func decryptData(_ data: Data, key: UnsafePointer<UInt8>) -> Data {
        return self.aesOperation(CCOperation(kCCDecrypt), on: data, key: key)
    }

    class func aesOperation(
        _ op: CCOperation,
        on data: Data,
        key: UnsafePointer<UInt8>
    ) -> Data {
        var buffer = [UInt8](repeating: 0, count: data.count + kCCKeySizeAES128)
        var bufferLen: size_t = 0
        CCCrypt(
            op,
            
            CCAlgorithm(kCCAlgorithmAES128),
            CCOptions(kCCOptionPKCS7Padding),
            key,
            kCCKeySizeAES128,
            nil,
            (data as NSData).bytes,
            data.count,
            &buffer,
            MemoryLayout.size(ofValue: buffer),
            &bufferLen)
        return Data(bytes: buffer, count: bufferLen)
    }

    class func encryptString(_ string: String) -> String {
        let data = string.data(using: .utf8)
        var dec: Data? = nil
        if let data = data {
            dec = aes.encryptData(data, key: key)
        }
        return dec?.base64EncodedString(options: []) ?? ""
    }

    class func decryptString(_ string: String) -> String {
        let data = Data(base64Encoded: string, options: [])
        var dec: Data? = nil
        if let data = data {
            dec = aes.decryptData(data, key: key)
        }
        if let dec = dec {
            return String(data: dec, encoding: .utf8) ?? ""
        }
        return ""
    }
}
